<launch>
  <!--
       Required arguments:
          HAND:=right|left
          TYPE:=A|B
       Suggested arguments:
          NUM:=0|1|...
          CONTROLLER:=grasp|pd|velsat|torque|sim
          KEYBOARD:=true|false  (default is true)
          AUTO_CAN:=true|false  (if true, ignores CAN_DEVICE argument and finds the can device automagically).
          CAN_DEVICE:=/dev/pcanusb1 | /dev/pcanusbNNN  (ls -l /dev/pcan* to see open CAN devices)
          VISUALIZE:=true|false  (Launch rviz)
	     MOVEIT:=true|false
	     GUI:=true|false
	  
       This script launches the following nodes:
         - allegro hand controller (different controllers exist)
         - keyboard controller
         - state publisher (for TF information)

       NOTE: If you specify the can device manually (CAN_DEVICE:=/dev/pcanusbN),
             make sure you *also* specify AUTO_CAN:=false.

  -->


  <!-- The inclusion of which_hand in the zero.yaml file has been deprecated.
       Which hand (left/right) must now be specified as an argument when launching the Allegro Hand
       as an input for both the robot_description and for the grasping library controllers. -->
  <arg name="HAND"/>
  <arg name="TYPE"/>
  <arg name="NUM" default="0"/>

  <!-- Use the joint_state_publisher for *commanded* joint angles. -->

  <!-- Visualization with rviz, only if arg VISUALIZE is set to true. Default is
       false, the allegro_viz.launch can be started separated. -->
  <arg name ="VISUALIZE" default="false" />
  <include file="$(find allegro_hand_controllers)/launch/allegro_viz.launch"
           if="$(arg VISUALIZE)">
    <arg name="NUM" value="$(arg NUM)"/>
     </include> 
     
  <arg name ="MOVEIT" default="false" />

  <include file="$(find allegro_hand_moveit)/launch/demo.launch" if="$(arg MOVEIT)">
  <arg name = "TYPE" value="$(arg TYPE)"/>
  <arg name = "HAND" value="$(arg HAND)"/>
  <arg name="NUM" value="$(arg NUM)"/>
  </include>

    
  <arg name="CONTROLLER" default="grasp"/> <!-- grasp, pd, velsat  -->
  <!-- Controllers include:
       -grasp *
       -pd
       -velsat
       -torque
       *The default controller is 'grasp' which employs the included grasping library (libBHand). -->

  <arg name="POLLING" default="true"/> <!-- true, false for polling the CAN communication -->

  <!-- ls -l /dev/pcan* to see your open CAN ports. Auto means find it
       automatically, and ignores the CAN_DEVICE argument. -->
  <arg name="AUTO_CAN" default="true" />
  <arg name="CAN_DEVICE" default="/dev/pcanusb1" />

  <arg name="PARAMS_DIR" default="$(find allegro_hand_parameters)" />
  <arg name="KEYBOARD" default="false" />

  <!-- yaml param files for your hand can be found in parameters/zero_files/ -->

  <!--If true, respawn the controller if it dies. -->
  <arg name="RESPAWN" default="false"/>

  <!-- Load the robot description directly from the xacro file. (NOTE: store it
       in two parameter names.) -->

   <param name="robot_description"
          command="cat $(find allegro_hand_description)/allegro_hand_description_$(arg HAND)_$(arg TYPE).urdf"/> 


  <!-- Allegro Hand controller and communication node. -->
  <node name="allegroHand_$(arg HAND)_$(arg NUM)"
        pkg="allegro_hand_controllers"
        type="allegro_node_$(arg CONTROLLER)"
        output="screen"
        clear_params="true"
        respawn="$(arg RESPAWN)"
        respawn_delay="2"
        args="$(arg POLLING)" >

    <!-- Remapping of topics into enumerated allegroHand_# namespace -->
    <remap from="allegroHand/joint_states" to="allegroHand_$(arg NUM)/joint_states"/>
    <remap from="allegroHand/joint_cmd" to="allegroHand_$(arg NUM)/joint_cmd"/>
    <remap from="allegroHand/lib_cmd" to="allegroHand_$(arg NUM)/lib_cmd"/>
    <remap from="allegroHand/torque_cmd" to="allegroHand_$(arg NUM)/torque_cmd"/>
    <remap from="allegroHand/envelop_torque" to="allegroHand_$(arg NUM)/envelop_torque"/>
    <remap from="allegroHand/joint_current_states" to="allegroHand_$(arg NUM)/joint_current_states"/>
    <remap from="allegroHand/joint_desired_states" to="allegroHand_$(arg NUM)/joint_desired_states"/>
    <remap from="/timechange" to="allegroHand_$(arg NUM)/timechange"/>
    <remap from="/forcechange" to="allegroHand_$(arg NUM)/forcechange"/>
    <remap from="fingertip_sensors" to="allegroHand_$(arg NUM)/fingertip_sensors"/>
    <remap from="fingertip_arrow_markers" to="allegroHand_$(arg NUM)/fingertip_arrow_markers"/>        

    <!--parameters are within the scope of the hand node so that multiple hands can be run at the same time -->


    <!-- Set the CAN channel automatically (using detect_pcan.py) if the
         AUTO_CAN parameter is true, otherwise use the CAN_DEVICE
         argument. NOTE: To manually set the can device, you must *also* set
         AUTO_CAN:=false. -->
    <param name="/comm/CAN_CH" value="$(arg CAN_DEVICE)"
           unless="$(arg AUTO_CAN)" />
    <param name="/comm/CAN_CH"
           command="$(find allegro_hand_description)/scripts/detect_pcan.py"
           if="$(arg AUTO_CAN)" />

    <param name="/hand_info/which_hand" value="$(arg HAND)" /> <!-- See HAND arg above -->
    <param name="/hand_info/which_type" value="$(arg TYPE)" />
  </node>

  <!-- Joint States (angles) to Joint Transforms -->
  <node name="jointState2tf_$(arg NUM)"
        pkg="robot_state_publisher"
        output="screen" 
        type="robot_state_publisher">
    <remap from="tf" to="allegroHand_$(arg NUM)/tf"/>
    <remap from="joint_states" to="allegroHand_$(arg NUM)/joint_states"/>
  </node>



  <!-- Keyboard handler (if arg KEYBOARD is true) -->
 
  <node name="keyboard_$(arg NUM)"
        pkg="allegro_hand_keyboard"
        type="allegro_hand_keyboard"
        output="screen"
        if="$(arg KEYBOARD)">
    <remap from="allegroHand/lib_cmd" to="allegroHand_$(arg NUM)/lib_cmd"/>
  </node>

  <arg name ="GUI" default="false" />
  <node name="allegro_hand_gui_node"
        pkg="allegro_hand_gui"
        type="allegro_hand_gui_node"
        output="screen"
        if="$(arg GUI)">
  <remap from="allegroHand/lib_cmd" to="allegroHand_$(arg NUM)/lib_cmd"/>
  <remap from="allegroHand/joint_cmd" to="allegroHand_$(arg NUM)/joint_cmd"/>
  <remap from="/timechange" to="allegroHand_$(arg NUM)/timechange"/>
  <remap from="/forcechange" to="allegroHand_$(arg NUM)/forcechange"/>
  </node>

  <!-- QT TEST(전시회) (if arg GUI is true) -->
<arg name ="DEMO" default="false" />

  <node name="rviz_embedded_node_$(arg NUM)"
        pkg="rviz_embedded" type="rviz_embedded_node"
        args="-d $(find allegro_hand_description)/allegro_hand_qt_config.rviz" 
        if="$(arg DEMO)">
    <remap from="/tf" to="/allegroHand_$(arg NUM)/tf"/>
    <remap from="fingertip_sensors" to="allegroHand_$(arg NUM)/fingertip_sensors"/>
     <remap from="/finger_labels" to="allegroHand_$(arg NUM)/finger_labels"/>
     <remap from="fingertip_arrow_markers" to="allegroHand_$(arg NUM)/fingertip_arrow_markers"/>  
    
  </node>
<arg name ="MISSION" default="false" />

  <node name="mission_node_$(arg NUM)"
        pkg="allegro_hand_mission"
        type="mission_node"
        output="screen"
        if="$(arg MISSION)">
    <remap from="allegroHand/lib_cmd" to="allegroHand_$(arg NUM)/lib_cmd"/>
    <remap from="allegroHand/forcechange" to="allegroHand_$(arg NUM)/forcechange"/>
</node>

</launch>
